### ISSUE TITLE
IOError on Windows

### ISSUE THREAD

**Issue by kke** [piotrmurach/tty-prompt#25]
Getting this:

```
`sysread': sysread for buffered IO (IOError)
from C:/Ruby/lib/ruby/gems/2.0.0/gems/tty-prompt-0.7.1/lib/tty/prompt/reader.rb:93
from C:/Ruby/lib/ruby/gems/2.0.0/gems/tty-prompt-0.7.1/lib/tty/prompt/reader.rb:78:in `block (3 levels) in read_keypress'
```

In reference to https://github.com/kontena/kontena/issues/1401

**Comment by kke** [comment]
0.7.2 would be nice :)

**Comment by piotrmurach** [comment]
I will have to release `v0.8.0` as I have fixed the behaviour for parsing individual keypresses. I think when bug fixes change functionality they deserve a major bump. Would that impact your release?

**Comment by kke** [comment]
Well, we have a big v1.0.0 release tomorrow of our product and 0.7.1 seems to be broken on windows..

**Comment by piotrmurach** [comment]
Give me few minutes and I will release `v0.8.0` for you to try out. Are you able to access windows box to test out?

**Comment by kke** [comment]
Hmm yes actually I do since I'm at home now.

**Comment by piotrmurach** [comment]
Could you actually test master branch to save us releasing unnecessary versions?

**Comment by kke** [comment]
Maybe. Going to take a while to install git, etc to my gaming pc.

Already tried $stdin.sysread(1) but can't get that to break, need to try with kontena cli.

**Comment by kke** [comment]
The menu navigation does not work at all for me on windows on 0.7.1.

![image](https://cloud.githubusercontent.com/assets/224971/20685725/3c06814e-b5be-11e6-9e4b-78017acc37f6.png)

Pressing arrows has no visual effect. Apparently the keypresses go through, but I can't see anything happening on screen.

When I press space or enter, it makes a choice.

I haven't seen the IOError yet.

**Comment by kke** [comment]
Also:

```ruby
prompt.yes?("Sure?")
prompt.ask("Talk to me:")
```

Pressing Y + enter on the first Y/n prompt leaks the enter to the second one every time and I don't get to answer the second one.

**Comment by piotrmurach** [comment]
Ok, to make sure this has nothing to do with any other dependencies, can you run the `examples/select` in the root project directory. I have added additional vim navigation keys `j` and `k` respectively. Can you check how the arrows work and whether vim like keys work. It maybe that ansi codes for up and down are not supported on windows(they should be on windows 10). 

If that's the case we need to find alternatives, running `examples/keypress` and actually pressing arrow up key should show what is actually detected. Can you try that as well and let me know the actual char codes. Here are [current codes](https://github.com/piotrmurach/tty-prompt/blob/master/lib/tty/prompt/reader/codes.rb)

**Comment by kke** [comment]
J+K work. Cleaning up doesn't.

![image](https://cloud.githubusercontent.com/assets/224971/20686827/d09b236a-b5c2-11e6-8ad8-8b2311aa3391.png)

I get nothing from arrow keys with keypress.rb.

Have you seen this: https://gist.github.com/acook/4190379

I get nothing from arrow keys on that either, plus the `STDIN.echo = false` gives Bad filedescriptor, but apart from that it seems to respond faster than tty-prompt.

**Comment by kke** [comment]
There's something like this done on some attempts to read keys in windows:

```ruby
    # save previous state of stty
    old_state = `stty -g`
    # disable echoing and enable raw (not having to press enter)
    system "stty raw -echo"

    ....

    # restore previous state of stty
    system "stty #{old_state}"
```

**Comment by piotrmurach** [comment]
Well the gist you found, reading comments, doesn't work on Windows. Using `stty` which is a typical Unix tool won't get you far on Windows either. I'm essentially switching mode through Ruby which as one of the fallbacks under covers uses `stty`. I think in order to simulate real raw input on windows you would need to use native Windows bindings or Cygwin. What windows version are you testing with?

**Comment by piotrmurach** [comment]
The clearing probably doesn't work because there is no ANSI codes support on the windows you're using which makes using this library impossible.

**Comment by piotrmurach** [comment]
Probably people on Windows would need to use [ansicon](https://github.com/adoxa/ansicon)

**Comment by kke** [comment]
I'm on Windows 10. Seems like remedy has pretty much the same problems. https://github.com/acook/remedy/issues/13 

Maybe just need to implement some "dumb" mode for windows.

**Comment by piotrmurach** [comment]
Yeah, I think I would need to take some time to investigate proper support on windows, maybe through `Win32API` etc... Have you considered [enum_select](https://github.com/piotrmurach/tty-prompt#272-enum_select) which doesn't require any arrow keys. It should work on windows I hope.

**Comment by kke** [comment]
Yes, maybe going with that. However the problem with two consequent questions not working (second one getting the enter from the first) is a pretty bad one.

**Comment by piotrmurach** [comment]
Ah I think I know why `select` doesn't work on Windows 10. You have to have `ENABLE_VIRTUAL_TERMINAL_PROCESSING` flag set in order to ANSI codes to work and the arrows being detect see https://msdn.microsoft.com/en-us/library/mt638032%28v=vs.85%29.aspx 
This flag is not enabled by default hence why `tty-prompt` detects enter correctly but fails to detect arrow up.

I will look into the other issue of subsequent inputs, it seem again windows specific.

**Comment by kke** [comment]
Everything works great on [Windows Subsystem for Linux](http://www.howtogeek.com/249966/how-to-install-and-use-the-linux-bash-shell-on-windows-10/), that's pretty ok workaround for my case.

Installation takes just a few clicks and a reboot and then I have apt-get et al, and tty-prompt works 100%.

**Comment by piotrmurach** [comment]
That's decent workaround. Given time I will install windows 10 image and play with bash a little bit there to see what's the crack. 

I will release new version then and going forward please report anything wrong with this library or even better contribute some code. I will do my best to fix things.

Shall we close this ticket for now as the original issue has been solved of IOError?

**Comment by piotrmurach** [comment]
Finally released `v0.8.0`.

**Comment by piotrmurach** [comment]
Closed as the original problem is fixed and I want to keep a single issue related to discussions about Windows support.
